File: Graphs/Medium/LC_200-Number_of_Islands/LC_200-Number_of_Islands.cpp
#include <bits/stdc++.h>

using namespace std;

// T & M: O(m * n), where m = rows and n = cols from grid input
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int rows = grid.size(), cols = grid[0].size();

        int count = 0;
        for (int r = 0; r < rows; ++r) {
            for (int c = 0; c < cols; ++c) {
                // If the current cell is land & not visited
                if (grid[r][c] == '1') {
                    dfs(r, c, rows, cols, grid);
                    // bfs(r, c, rows, cols, grid);
                    ++count;
                }
            }
        }
        return count;
    }

private:  
    void dfs(int r, int c, int rows, int cols, vector<vector<char>>& grid) {
        if (r < 0 || rows <= r ||
        c < 0 || cols <= c ||
        grid[r][c] != '1')
            return; // water reached or cell already visited

        grid[r][c] = 'V';   // cell visited

        // Recursively check neighboring cells
        dfs(r+1, c, rows, cols, grid);
        dfs(r-1, c, rows, cols, grid);
        dfs(r, c+1, rows, cols, grid);
        dfs(r, c-1, rows, cols, grid);
    }
    

    /*
    struct indicies {const int row, col;};

    void bfs(int r, int c, int rows, int cols, vector<vector<char>>& grid) {
        queue<indicies> q;
        grid[r][c] = 'V';
        q.push({r, c});

        // while there are connected land cells to visit
        while (!q.empty()) {
            indicies currCell = q.front();
            q.pop();	// visit cells in FCFS/FIFO order
            vector<indicies> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
            for (indicies &dir : directions) {
                int neighbR = currCell.row + dir.row;
                int neighbC = currCell.col + dir.col;
                // if neigboring cells are land & not visited
                if ((-1 < neighbR && neighbR < rows) &&
                (-1 < neighbC && neighbC < cols) &&
                grid[neighbR][neighbC] == '1') {
                    q.push({neighbR, neighbC});
                    grid[neighbR][neighbC] = 'V';
                }
            }
        }
    }
    */

/*
// BFS Solution with set of visited coordinates (object(s))
private:
    vector<vector<char>> grid;
    int rows, cols;
    struct coordinates {const int row, col;};
    struct coordinates_hash {
        size_t operator()(const coordinates &coord) const {
            return hash<int>()(coord.row) ^ hash<int>()(coord.col);
        }
    };
    struct coordinates_equal {
        bool operator()(const coordinates &lhs, const coordinates &rhs) const {
            return lhs.row == rhs.row && lhs.col == rhs.col;
        }
    };
    unordered_set<coordinates, coordinates_hash, coordinates_equal> visited;

    void bfs(int r, int c) {
        queue<coordinates> q;
        visited.insert({r, c});
        q.push({r, c});

        // while there are connected land coordinates to visit
        while (!q.empty()) {
            coordinates currCoords = q.front();
            q.pop();	// visit coordinates in FCFS/FIFO order
            vector<coordinates> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
            for (coordinates &dir : directions) {
                int neighbR = currCoords.row + dir.row;
                int neighbC = currCoords.col + dir.col;
                // if neigboring coordinates are land & not visited
                if ((-1 < neighbR && neighbR < rows) &&
                (-1 < neighbC && neighbC < cols) &&
                grid[neighbR][neighbC] == '1' &&
                visited.find({neighbR, neighbC}) == visited.end()) {
                    q.push({neighbR, neighbC});
                    visited.insert({neighbR, neighbC});
                }
            }
        }
    }
*/
};

int main() {
    int attempt;
    vector<vector<char>> grid;

    // Example 1
    grid = {
        {'1','1','1','1','0'},
        {'1','1','0','1','0'},
        {'1','1','0','0','0'},
        {'0','0','0','0','0'}
    };
    attempt = Solution().numIslands(grid);
    assert(attempt == 1 && "Failed Example 1: Expected 1 island");

    // Example 2
    grid = {
        {'1','1','0','0','0'},
        {'1','1','0','0','0'},
        {'0','0','1','0','0'},
        {'0','0','0','1','1'}
    };
    attempt = Solution().numIslands(grid);
    assert(attempt == 3 && "Failed Example 2: Expected 3 islands");

    // Test Case 3:
    grid = {
        {'0','0','1','0','0','1','0','0','0','0','0','0','0','0','1','1','0','1','0','0'},
        {'0','1','0','0','0','0','1','1','0','0','0','1','0','0','1','1','0','0','0','0'},
        {'1','0','1','1','0','0','0','0','0','1','0','0','0','1','0','1','1','1','1','0'},
        {'1','1','0','0','0','0','0','0','0','0','1','0','0','1','1','1','1','1','1','1'},
        {'0','0','0','0','0','1','0','0','0','1','1','1','1','0','1','0','0','0','0','0'},
        {'0','1','1','1','0','0','0','1','0','1','0','1','0','0','1','0','1','1','0','0'},
        {'0','0','0','0','0','0','0','0','1','1','1','0','0','1','1','0','0','0','0','0'},
        {'0','1','1','0','0','0','0','0','1','0','1','1','0','1','1','0','0','1','0','0'},
        {'0','0','1','1','1','0','0','1','0','0','0','0','0','0','0','1','1','1','0','1'},
        {'1','1','0','0','0','1','0','1','0','0','0','1','1','0','0','1','0','1','1','0'},
        {'0','0','0','0','0','0','1','0','1','1','0','0','1','0','1','1','1','1','0','1'},
        {'0','0','1','1','0','0','1','0','1','0','0','1','0','0','1','0','0','1','0','1'},
        {'0','1','0','0','0','0','0','1','0','0','0','0','0','0','1','1','1','0','0','0'},
        {'0','0','1','0','1','0','0','1','1','0','1','1','1','0','0','1','1','0','0','1'},
        {'1','0','1','0','1','0','1','0','0','0','0','0','0','0','1','0','0','1','1','0'},
        {'1','0','1','1','1','0','1','0','0','0','0','0','0','1','0','0','0','0','1','1'},
        {'1','0','0','0','0','1','0','0','0','0','0','0','1','0','0','0','0','0','1','1'},
        {'0','0','0','0','0','0','1','0','0','0','1','0','1','0','1','1','0','1','0','1'},
        {'1','0','0','1','0','0','0','0','0','0','1','0','0','0','0','0','1','1','1','1'},
        {'0','0','0','1','1','0','0','0','0','0','0','0','0','1','0','0','0','1','0','0'}
    };
    attempt = Solution().numIslands(grid);
    assert(attempt == 45 && "Failed Test Case 3: Expected 45 islands");

    return 0;
}

File: Graphs/Medium/LC_200-Number_of_Islands/LC_200-Number_of_Islands.py
import collections

# T & M: O(m * n), where m = rows and n = cols from grid input
class Solution:
    def numIslands(self, grid: list[list[str]]) -> int:
        self.grid = grid
        self.rows, self.cols = len(grid), len(grid[0])
        self.visited = set()

        count = 0
        for r in range(self.rows):
            for c in range(self.cols):
                # if land and coordinates not visited
                if grid[r][c] == "1" and (r, c) not in self.visited:
                    self.dfs(r, c)
                    # self.bfs(r, c)
                    count += 1
        return count

    def dfs(self, r: int, c: int) -> None:
        if (
            r < 0 or self.rows <= r or
            c < 0 or self.cols <= c or
            self.grid[r][c] == "0" or
            (r, c) in self.visited 
        ):
            return  # water reached or coordinates already visited
        
        self.visited.add((r, c))
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        for dir_r, dir_c in directions:
            self.dfs(r + dir_r, c + dir_c)  # go deeper into path

    def bfs(self, r: int, c: int) -> None:
        self.visited.add((r, c))
        q = collections.deque() # visit coordinates in FCFS/FIFO order
        q.append((r, c))

        # while there are connected land coordinates to visit
        while q:
            curr_r, curr_c = q.popleft()
            directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
            for dir_r, dir_c in directions:
                neighb_r, neighb_c = curr_r + dir_r, curr_c + dir_c
                # if neigboring coordinates are land & not visited
                if (
                    -1 < neighb_r < self.rows and
                    -1 < neighb_c < self.cols and
                    self.grid[neighb_r][neighb_c] == "1" and
                    (neighb_r, neighb_c) not in self.visited
                ):
                    q.append((neighb_r, neighb_c))
                    self.visited.add((neighb_r, neighb_c))
    
sol = Solution()

# Example 1:
attempt = sol.numIslands(grid = [
    ["1","1","1","1","0"],
    ["1","1","0","1","0"],
    ["1","1","0","0","0"],
    ["0","0","0","0","0"]
])
assert attempt == 1, f'Expected 1, but got {attempt}'

# Example 2:
attempt = sol.numIslands(grid = [
    ["1","1","0","0","0"],
    ["1","1","0","0","0"],
    ["0","0","1","0","0"],
    ["0","0","0","1","1"]
])
assert attempt == 3, f'Expected 3, but got {attempt}'

# Test Case 3:
attempt = sol.numIslands(grid = [
    ["0","0","1","0","0","1","0","0","0","0","0","0","0","0","1","1","0","1","0","0"],
    ["0","1","0","0","0","0","1","1","0","0","0","1","0","0","1","1","0","0","0","0"],
    ["1","0","1","1","0","0","0","0","0","1","0","0","0","1","0","1","1","1","1","0"],
    ["1","1","0","0","0","0","0","0","0","0","1","0","0","1","1","1","1","1","1","1"],
    ["0","0","0","0","0","1","0","0","0","1","1","1","1","0","1","0","0","0","0","0"],
    ["0","1","1","1","0","0","0","1","0","1","0","1","0","0","1","0","1","1","0","0"],
    ["0","0","0","0","0","0","0","0","1","1","1","0","0","1","1","0","0","0","0","0"],
    ["0","1","1","0","0","0","0","0","1","0","1","1","0","1","1","0","0","1","0","0"],
    ["0","0","1","1","1","0","0","1","0","0","0","0","0","0","0","1","1","1","0","1"],
    ["1","1","0","0","0","1","0","1","0","0","0","1","1","0","0","1","0","1","1","0"],
    ["0","0","0","0","0","0","1","0","1","1","0","0","1","0","1","1","1","1","0","1"],
    ["0","0","1","1","0","0","1","0","1","0","0","1","0","0","1","0","0","1","0","1"],
    ["0","1","0","0","0","0","0","1","0","0","0","0","0","0","1","1","1","0","0","0"],
    ["0","0","1","0","1","0","0","1","1","0","1","1","1","0","0","1","1","0","0","1"],
    ["1","0","1","0","1","0","1","0","0","0","0","0","0","0","1","0","0","1","1","0"],
    ["1","0","1","1","1","0","1","0","0","0","0","0","0","1","0","0","0","0","1","1"],
    ["1","0","0","0","0","1","0","0","0","0","0","0","1","0","0","0","0","0","1","1"],
    ["0","0","0","0","0","0","1","0","0","0","1","0","1","0","1","1","0","1","0","1"],
    ["1","0","0","1","0","0","0","0","0","0","1","0","0","0","0","0","1","1","1","1"],
    ["0","0","0","1","1","0","0","0","0","0","0","0","0","1","0","0","0","1","0","0"]
])
assert attempt == 45, f'Expected 45, but got {attempt}'